### ISSUE TITLE
Reproducibility of Estimators

### ISSUE THREAD

**Issue by knoam** [microsoft/FLAML#236]
Is there a way to pass parameters to the estimators to make them reproducible? For example, LightGBM's `seed` and `deterministic` arguments?

**Comment by sonichi** [comment]
You can add a custom learner class by inheriting `flaml.model.LGBMEstimator` and set these parameters in `self.params` in its constructor.

**Comment by knoam** [comment]
Thanks @sonichi. In your opinion, would using the same random seed for LightGBM during training impact generalization? One could also increment the random seed at each call, something like this:

```python
global random_state
random_state = 0
class LGBMEstimatorSeeded(LGBMEstimator):
    def __init__(self, task='binary', **config):
        
        super().__init__(task, **config)
        
        global random_state
        random_state += 1
        self.params.update(dict(
            deterministic=True,
            force_row_wise=True,  # see https://lightgbm.readthedocs.io/en/latest/Parameters.html#deterministic
            seed=random_state, 
        ))
```

**Comment by knoam** [comment]
Thanks for your response. One issue with the solution above is that the number of iterations can change from run to run, so reproducibility is not guaranteed.

**Comment by sonichi** [comment]
@knoam Could you share more about your use case? Depending on your goal, it may be ok to use a fixed random seed.

**Comment by knoam** [comment]
@sonichi I work in a regulated industry where everything has to be reproducible for verification purposes. This includes hyperparameter tuning. So if an auditor were to repeat the tuning, they would need to be able to arrive at the same hyperparameters.

**Comment by sonichi** [comment]
@knoam In that case I would suggest using a fixed random seed. That is already the case by default. In addition, you can set time_budget=None and max_iter to be a constant number. Then you are supposed to get the same result every time when tuning lightgbm on the same dataset. If that's not the case, please let us know.

**Comment by knoam** [comment]
I am having trouble with reproducibility even after following your advice above. Here's a reproducible example:

```python
from flaml.data import load_openml_dataset
from flaml.ml import ExtraTreesEstimator
from flaml import AutoML

X_train, X_test, y_train, y_test = load_openml_dataset(dataset_id=1169, data_dir='./')
X_train = X_train.iloc[:1000]
y_train = y_train.iloc[:1000]

class ExtraTreesEstimatorSeeded(ExtraTreesEstimator):
    def __init__(self, task='binary', **config):
        'ExtraTreesEstimator for reproducible FLAML run'
        super().__init__(task, **config)
        self.params.update({'random_state':0})

settings = {
    "time_budget": 1e10,   # total running time in seconds
    "max_iter": 3,   
    "metric": 'ap',        # average_precision
    "task": 'classification',  # task type
    "seed": 0,    # random seed
    "estimator_list": ['extra_trees_seeded'],
    "verbose": False
}

for trial_num in range(8):
    automl = AutoML()
    automl.add_learner(learner_name='extra_trees_seeded', learner_class=ExtraTreesEstimator)
    automl.fit(X_train=X_train, y_train=y_train, **settings)
    print(automl.best_loss)
    print(automl.best_config)
```

And here is the output:

```python
load dataset from ./openml_ds1169.pkl
Dataset name: airlines
X_train.shape: (404537, 7), y_train.shape: (404537,);
X_test.shape: (134846, 7), y_test.shape: (134846,)
0.3979896835033882
{'n_estimators': 6, 'max_features': 0.5135217307583296, 'max_leaves': 4, 'criterion': 'gini'}
0.3899486634000108
{'n_estimators': 6, 'max_features': 0.5135217307583296, 'max_leaves': 4, 'criterion': 'gini'}
0.3899178957824727
{'n_estimators': 6, 'max_features': 0.5135217307583296, 'max_leaves': 4, 'criterion': 'entropy'}
0.3899178957824727
{'n_estimators': 6, 'max_features': 0.5135217307583296, 'max_leaves': 4, 'criterion': 'entropy'}
0.3899178957824727
{'n_estimators': 6, 'max_features': 0.5135217307583296, 'max_leaves': 4, 'criterion': 'entropy'}
0.3899178957824727
{'n_estimators': 6, 'max_features': 0.5135217307583296, 'max_leaves': 4, 'criterion': 'entropy'}
0.3899486634000108
{'n_estimators': 6, 'max_features': 0.5135217307583296, 'max_leaves': 4, 'criterion': 'gini'}
0.3899486634000108
{'n_estimators': 6, 'max_features': 0.5135217307583296, 'max_leaves': 4, 'criterion': 'gini'}
```

**Comment by sonichi** [comment]
@knoam the problem is fixed in #349 
could you try installing the latest version from git:
`pip install git+https://github.com/microsoft/FLAML.git`
and check if your problem is solved?
I've tested [test/rep.py](https://github.com/microsoft/FLAML/blob/main/test/rep.py)

**Comment by knoam** [comment]
I confirm that now extra_tree model (using ExtraTreesEstimatorSeeded as described above) is reproducible for my dataset. Thanks!
